# 知识点

1. <font color=red>**简述GC**</font>

   1. 标记清除 v1.3

      > - 根对象出发寻找并标记可达对象
      > - 遍历堆中的对象，将未标记的对象回收
      >
      > 以上步骤都需要暂停程序：STW； 扫描整个堆；

   2. 三色标记 v.15

      > _黑白灰三色_
      >
      > - 最开始全部放入白色标记表
      > - 扫表根节点，可达对象标记为灰色，一次遍历之后当前灰色根节点置为黑色
      > - 重复，直到灰色表无数据
      > - 白色清除
      >
      > 在标记过程中还是需要STW，如果没有STW会有两次场景下gc问题
      >
      > - 在gc时，黑色链接白色
      > - 在gc时，灰色断开白色
      >
      > 针对上面两种情况，有了下面两个公式以及插入写、删除写屏障

   3. 三色标记+混合写屏障v1.8

      > 强三色：黑色不能直接引用白色
      >
      > 弱三色：黑色引用白色对象处于灰色保护状态
      >
      > 应用到下面：
      >
      > 插入写：**黑引用白，则白色置为灰色(仅在堆中生效)  栈上则是全部重置白色，STW开启 重新扫描一遍**
      >
      > 删除写：灰引用白色，断开时，白色变为灰色(弱三色？) 下次gc才会被删除

2. make和new的区别

> 1.make适用slice、map、chan；new适用基本类型；
> 2.make返回引用，new返回地址
> 3.make初始空间，new为零值，并回收内存空间
3. defer和return执行顺序
> defer在return执行后面
> defer压栈原理，后进先出
> defer拷贝传值，不会影响return值； 如果地址传参，则会影响return值
4. print
> 1. %v(只打印结构体val)
> 2. %+v(打印结构体key/val)
> 3. %#v(打印结构体命名空间、key/val)
```go
func TestStruct(t *testing.T) {
	type Ts struct {
		Name string `json:"name"`
		Age  int    `json:"age"`
	}

	obj := Ts{
		Name: "哈哈怪",
		Age:  18,
	}
	fmt.Printf("%v\n", obj)
	fmt.Printf("%+v\n", obj)
	fmt.Printf("%#v\n", obj)
}

```
Output:
```
{哈哈怪 18}
{Name:哈哈怪 Age:18}
main.Ts{Name:"哈哈怪", Age:18}
PASS
```
5. 空结构体的作用
> 1. 不占任何空间
> 2. 拓展普通结构体时的内存对齐， 小的在前面排列
> [参考链接](https://juejin.cn/post/7244809769794207801)

6. for循环里 传递value引用 将会全是最后一个元素内容

```go
// 这里不能为了考虑节省空间而使用指针
for _, value := range sliceList {
    go Get(&value) // value全是sliceList的最后一个元素内容
}
应该是
for _, value := range sliceList {
    Get(value) 
}

或

for _, value := range sliceList {
    value := value
    go Get(&value) 
}
```

# 并发

1. 有无缓冲channel区别
2. 

## 协程中的panic
panic触发只会中断当前协程的执行，不会影响其他协程，所以如果在为了能捕获到协程中的panic错误，我们应该用recover()进行捕捉处理, 发现问题，解决问题
例如:
```go
func doSomeThing() error {
	fmt.Println("do some thing...")
	panic("test")
	return nil
}

func main() {
	for i := 10; i > 0; i-- {
		go func() {
			if err := doSomeThing(); err != nil {
				fmt.Println(err.Error())
			}
		}()
	}
	time.Sleep(1)
}
```
这里创建的子协程中的panic不会被发现, 我们需要用recover进行捕获
```go
package main

import (
	"fmt"
	"time"

	"github.com/spf13/cast"
)

func doSomeThing() error {
	defer func() {
		if err := recover(); err != nil {
			fmt.Println(cast.ToString(err))
		}
	}()
	fmt.Println("do some thing...")
	panic("test")

	return nil
}

func main() {
	for i := 10; i > 0; i-- {
		go func() {
			if err := doSomeThing(); err != nil {
				fmt.Println("doSomeThing err:", err.Error())
			}
		}()
	}
	time.Sleep(1)
}
```
